WIP IO work

piatra opened the pull request.

Commit:
Bug 1668435 - add nimbus-sdk + uniffi refs and various glue

Differential Revision: https://phabricator.services.mozilla.com/D94412

Commit:
Bug 1668435 - Vendor in Nimbus SDK and friends

Differential Revision: https://phabricator.services.mozilla.com/D95392

Depends on D94412

Commit:
Bug 1668435 - Add Nimbus artifacts generated by uniffi branch

Differential Revision: https://phabricator.services.mozilla.com/D92538

Depends on D92537

Commit:
Bug 1668435 - Nimbus SDK build glue

Differential Revision: https://phabricator.services.mozilla.com/D92537

Depends on D95392

Commit:
Bug 1668435 - Add xpcshell port of test_fs_client.rs

Differential Revision: https://phabricator.services.mozilla.com/D92539

Depends on D92538

piatra committed:
Compiling

piatra committed:
Still crashing when trying to do a network request

dmose commented:
@piatra I rebased this branch on top of my branch so that it didn't end up having duplicated commits to land and build the nimbus-sdk.  You'll probably want to pull my branch also, and then forcibly merge this new version of your patches so that we're all in sync.

Commit:
Moved the webidl file to the dom directory, as per Nika.

Nika said that things were very likely to be broken in unspecified ways if the webidl
wasn't under the dom/ subtree.  So I moved this, following Glean's pattern to some degree.
Unfortunately, it doesn't seem to have fixed the crash.

Commit:
Fix crash from accessing NimbusClient off the main thread

Commit:
Crashes fixed and test passes; paired w/@brennie (thanks!).  But...

We probably don't want to go with this exact approach, because the NimbusClient participates in cycle collection,
and objects that do that can't implement a threadsafe nsISupports.  This patch really just hacks around it.
Barret has suggested another possible approach, and a few others have come up during other conversations as well...

Commit:
A different approach...

dmose commented:
@mystor was kind enough to help me work through a bunch of the existing code last Friday AM.  The conclusion was that it's not really possible to make stuff that participates in cycle collection threadsafe in the way that would make this sort of code truly robust, and she suggest a different approach, and even sketched out some code for me.

The intent is to end with something approximately similar to https://searchfox.org/mozilla-central/rev/44e6dfd7e02edd95e5fd4d4c25c8b946131f92cd/xpcom/base/nsSystemInfo.cpp#1334 and the code in that area.

I've spent a bunch of time hacking on this approach, and I've got something very basic compiling, but no matter what I do, the code in this PR now crashes on the call to InvokeAsync.  @mystor or @brennie, if either you see anything obvious or have any suggestions, @piatra and I are all ears...

Commit:
GetActiveExperimentsAsync now up and limping

Commit:
Misc cleanups

Commit:
Updated generated webidl from GeckoJS branch -- C++ code in the branch doesn't yet compile with this

Commit:
Updated generated C++ code from uniffi geckojs_async branch; doesn't yet compile

Commit:
Generated async bindings that now compile

Commit:
Update test to work with async bindings

dmose commented:
OK, this now generates async bindings for all the methods, and the getActiveExperiments passes, amazingly enough!

piatra committed:
Connect ExperimentManager to the NimbusSDK

piatra committed:
Fix bucket name and update store.get call

Commit:
Progress towards keying off the slug name of the branch

Commit:
Fix up some errors because the Rust Nimbus objects are shaped a bit differently

Commit:
Make GetActiveExperiments emit a more useful error -- should backport to uniffi

piatra commented on browser/actors/AboutNewTabChild.jsm:
```diff
@@ -39,6 +39,16 @@ class AboutNewTabChild extends JSWindowActorChild {
+        {slug: "multistage_aboutwelcome_test"}
+      );
+      if (
+        experimentData?.slug == "multistage_aboutwelcome_test_treatment" &&
```
I think this should say `experimentData?.branch.slug`?

piatra committed:
Working demo

piatra committed:
Small rebase fix

piatra committed:
Add cache busting url

piatra committed:
Use the right ARU client_id

piatra committed:
Read the user id from the prefs

dmose commented:
@markh Note that the last commit here which tries to read the randomization unit from the prefs doesn't actually work right, so while this branch is buildable and the unit test can be made to run, that last bit doesn't work as was hoped (prefs aren't available that early in startup, we think).
